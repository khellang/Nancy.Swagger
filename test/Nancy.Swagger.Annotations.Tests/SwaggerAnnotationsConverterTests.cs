using ApprovalTests;
using ApprovalTests.Reporters;
using Nancy.Swagger.Modules;
using Nancy.Swagger.Services;
using Nancy.Testing;
using System;
using System.Linq;
using Xunit;

namespace Nancy.Swagger.Annotations.Tests
{
    [UseReporter(typeof(XUnitReporter), typeof(DiffReporter))]
    public class SwaggerAnnotationsConverterTests
    {
        private readonly Browser _browser;

        public SwaggerAnnotationsConverterTests()
        {
            var bootstrapper = new ConfigurableBootstrapper(with =>
            {
                with.ApplicationStartup((container, pipelines) =>
                    container.Register<ISwaggerMetadataConverter, SwaggerAnnotationsConverter>());

                with.Module<SwaggerModule>();
                with.Module<TestRoutesModule>();
            });

            _browser = new Browser(bootstrapper);
        }

        [Fact]
        public void Get_ApiDocsPath_ReturnsApiDeclaration()
        {
            ApproveJsonResponse(_browser.Get("/api-docs/api-docs"));
        }

        [Fact]
        public void Get_ApiDocsRootpath_ReturnsResourceListing()
        {
            ApproveJsonResponse(_browser.Get("/api-docs"));
        }

        [Fact]
        public void Get_TestModulePath_ReturnsApiDeclaration()
        {
            ApproveJsonResponse(_browser.Get("/api-docs/testroutes"));
        }

        private static void ApproveJson(string json)
        {
            Approvals.Verify(FormatJson(json));
        }

        private static void ApproveJsonResponse(BrowserResponse response)
        {
            ApproveJson(response.Body.AsString());
        }

        private static string FormatJson(string json)
        {
            const string indentString = "    ";

            var indentation = 0;
            var quoteCount = 0;

            var result =
                from ch in json
                let quotes = ch == '"' ? quoteCount++ : quoteCount
                let lineBreak = ch == ',' && quotes % 2 == 0 ? ch + Environment.NewLine + String.Concat(Enumerable.Repeat(indentString, indentation)) : null
                let openChar = (ch == '{' || ch == '[') && quotes % 2 == 0 ? ch + Environment.NewLine + String.Concat(Enumerable.Repeat(indentString, ++indentation)) : ch.ToString()
                let closeChar = (ch == '}' || ch == ']') && quotes % 2 == 0 ? Environment.NewLine + String.Concat(Enumerable.Repeat(indentString, --indentation)) + ch : ch.ToString()
                select lineBreak ?? (openChar.Length > 1 ? openChar : closeChar);

            return String.Concat(result);
        }
    }
}