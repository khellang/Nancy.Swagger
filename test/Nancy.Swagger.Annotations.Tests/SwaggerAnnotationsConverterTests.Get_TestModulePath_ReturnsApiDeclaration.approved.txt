{
    "swaggerVersion":"1.2",
    "basePath":"/",
    "apis":[
        {
            "path":"/testroutes/allowmultipleparam",
            "operations":[
                {
                    "method":"GET",
                    "nickname":"/testroutes/allowmultipleparam",
                    "parameters":[
                        {
                            "paramType":"query",
                            "name":"ids",
                            "required":false,
                            "allowMultiple":true,
                            "type":"integer",
                            "format":"int32"
                        }
                    ],
                    "responseMessages":[
                        
                    ],
                    "produces":[
                        
                    ],
                    "consumes":[
                        
                    ],
                    "type":"void"
                }
            ]
        },
        {
            "path":"/testroutes/anonymoushandler",
            "operations":[
                {
                    "method":"GET",
                    "summary":"Warning: no annotated method found for this route",
                    "notes":"[example]",
                    "nickname":"/testroutes/anonymoushandler",
                    "parameters":[
                        
                    ],
                    "responseMessages":[
                        
                    ],
                    "produces":[
                        
                    ],
                    "consumes":[
                        
                    ],
                    "type":"void"
                }
            ]
        },
        {
            "path":"/testroutes/models",
            "operations":[
                {
                    "method":"GET",
                    "nickname":"/testroutes/models",
                    "parameters":[
                        
                    ],
                    "responseMessages":[
                        
                    ],
                    "produces":[
                        
                    ],
                    "consumes":[
                        
                    ],
                    "type":"array",
                    "items":{
                        "$ref":"TestModel"
                    }
                }
            ]
        },
        {
            "path":"/testroutes/models/{id}",
            "operations":[
                {
                    "method":"GET",
                    "nickname":"/testroutes/models/{id}",
                    "parameters":[
                        {
                            "paramType":"path",
                            "name":"id",
                            "required":true,
                            "allowMultiple":false,
                            "type":"integer",
                            "format":"int32"
                        }
                    ],
                    "responseMessages":[
                        
                    ],
                    "produces":[
                        
                    ],
                    "consumes":[
                        
                    ],
                    "type":"TestModel"
                },
                {
                    "method":"POST",
                    "nickname":"/testroutes/models/{id}",
                    "parameters":[
                        {
                            "paramType":"body",
                            "name":"body",
                            "required":true,
                            "type":"TestModel"
                        }
                    ],
                    "responseMessages":[
                        {
                            "code":200,
                            "message":"Everything OK",
                            "responseModel":"TestModel"
                        },
                        {
                            "code":404,
                            "message":"NotFound"
                        },
                        {
                            "code":418,
                            "message":"I'm a teapot"
                        },
                        {
                            "code":500,
                            "message":"InternalServerError",
                            "responseModel":"string"
                        }
                    ],
                    "produces":[
                        
                    ],
                    "consumes":[
                        
                    ],
                    "type":"void"
                },
                {
                    "method":"PUT",
                    "nickname":"/testroutes/models/{id}",
                    "parameters":[
                        {
                            "paramType":"body",
                            "name":"body",
                            "required":false,
                            "type":"TestModel"
                        }
                    ],
                    "responseMessages":[
                        
                    ],
                    "produces":[
                        
                    ],
                    "consumes":[
                        
                    ],
                    "type":"void"
                },
                {
                    "method":"PATCH",
                    "nickname":"/testroutes/models/{id}",
                    "parameters":[
                        {
                            "paramType":"body",
                            "name":"body",
                            "required":false,
                            "type":"TestModel"
                        }
                    ],
                    "responseMessages":[
                        
                    ],
                    "produces":[
                        
                    ],
                    "consumes":[
                        
                    ],
                    "type":"void"
                },
                {
                    "method":"DELETE",
                    "nickname":"/testroutes/models/{id}",
                    "parameters":[
                        {
                            "paramType":"path",
                            "name":"id",
                            "required":true,
                            "allowMultiple":false,
                            "type":"integer",
                            "format":"int32"
                        }
                    ],
                    "responseMessages":[
                        
                    ],
                    "produces":[
                        
                    ],
                    "consumes":[
                        
                    ],
                    "type":"void"
                },
                {
                    "method":"OPTIONS",
                    "nickname":"/testroutes/models/{id}",
                    "parameters":[
                        {
                            "paramType":"path",
                            "name":"id",
                            "required":true,
                            "allowMultiple":false,
                            "type":"integer",
                            "format":"int32"
                        }
                    ],
                    "responseMessages":[
                        
                    ],
                    "produces":[
                        
                    ],
                    "consumes":[
                        
                    ],
                    "type":"void"
                }
            ]
        },
        {
            "path":"/testroutes/namedroute",
            "operations":[
                {
                    "method":"GET",
                    "nickname":"/testroutes/namedroute",
                    "parameters":[
                        
                    ],
                    "responseMessages":[
                        
                    ],
                    "produces":[
                        
                    ],
                    "consumes":[
                        
                    ],
                    "type":"array",
                    "items":{
                        "type":"integer",
                        "format":"int32"
                    }
                }
            ]
        },
        {
            "path":"/testroutes/not-annotated/get",
            "operations":[
                {
                    "method":"GET",
                    "summary":"Warning: no annotated method found for this route",
                    "notes":"[example]",
                    "nickname":"/testroutes/not-annotated/get",
                    "parameters":[
                        
                    ],
                    "responseMessages":[
                        
                    ],
                    "produces":[
                        
                    ],
                    "consumes":[
                        
                    ],
                    "type":"void"
                }
            ]
        },
        {
            "path":"/testroutes/strings",
            "operations":[
                {
                    "method":"GET",
                    "summary":"Some summary",
                    "notes":"Some notes",
                    "nickname":"/testroutes/strings",
                    "parameters":[
                        
                    ],
                    "responseMessages":[
                        
                    ],
                    "produces":[
                        
                    ],
                    "consumes":[
                        
                    ],
                    "type":"array",
                    "items":{
                        "type":"string"
                    }
                }
            ]
        },
        {
            "path":"/testroutes/strings/{id}",
            "operations":[
                {
                    "method":"GET",
                    "nickname":"/testroutes/strings/{id}",
                    "parameters":[
                        {
                            "paramType":"path",
                            "name":"id",
                            "required":true,
                            "allowMultiple":false,
                            "type":"integer",
                            "format":"int32"
                        },
                        {
                            "paramType":"query",
                            "name":"q",
                            "description":"Query",
                            "required":false,
                            "allowMultiple":false,
                            "type":"string"
                        }
                    ],
                    "responseMessages":[
                        
                    ],
                    "produces":[
                        "application/json"
                    ],
                    "consumes":[
                        "application/json",
                        "application/xml"
                    ],
                    "type":"string"
                }
            ]
        }
    ],
    "models":{
        "TestModel":{
            "id":"TestModel",
            "description":"Description of a model",
            "required":[
                "ExplicitRequired",
                "ImplicitRequired"
            ],
            "properties":{
                "Description":{
                    "description":"Some description",
                    "type":"string"
                },
                "Enum":{
                    "type":"string",
                    "enum":[
                        "male",
                        "female"
                    ]
                },
                "ExplicitRequired":{
                    "type":"string"
                },
                "ImplicitRequired":{
                    "description":"Non-nullable value types are implicitly required",
                    "type":"integer",
                    "format":"int32"
                },
                "ListOfUniqueItems":{
                    "type":"array",
                    "items":{
                        "type":"string"
                    },
                    "uniqueItems":true
                },
                "Maximum":{
                    "type":"integer",
                    "format":"int32",
                    "maximum":100
                },
                "Minimum":{
                    "type":"integer",
                    "format":"int32",
                    "minimum":0
                },
                "MultipleAnnotations":{
                    "description":"Property with multiple annotations (minimum, maximum and description)",
                    "type":"integer",
                    "format":"int32",
                    "minimum":0,
                    "maximum":100
                },
                "name-by-constructor":{
                    "type":"string"
                },
                "name-by-namedparameter":{
                    "type":"string"
                }
            }
        }
    }
}